File: JAVASCRIPT/Algorithms/Sort/5. MergeSort.js
const generateArray = require('./ArrayGeneration');

// Generates an array of 500000 random numbers from -1000000 to 1000000
const arr500K = generateArray(500000, -1000000, 1000000);

const mergeHelper = (arr1, arr2) => {
  const result = [];

  let i = 0;
  let j = 0;

  while (i < arr1.length && j < arr2.length) {
    if (arr1[i] < arr2[j]) {
      result.push(arr1[i]);
      i++;
    } else {
      result.push(arr2[j]);
      j++;
    }
  }
  result.push(...arr1.slice(i));
  result.push(...arr2.slice(j));

  return result;
};

const mergeSort = (arr) => {
  if (arr.length < 2) {
    return arr;
  }

  const middle_index = arr.length / 2;
  const left = mergeSort(arr.slice(0, middle_index));
  const right = mergeSort(arr.slice(middle_index));

  return mergeHelper(left, right);
};

const arr = [...arr500K];

// Measure execution time
const startTime = performance.now();
const sortedArr = mergeSort(arr);
// if you want to visualize starting array just change sortedArr to arr500K
const sortedArrayString = sortedArr.join(', ');
const endTime = performance.now();

// It will take at least 0.2 seconds to sort the arr500K
const timeTakenInSeconds = (endTime - startTime) / 1000;

console.log('[', sortedArrayString, ']');
console.log('--------------------------------------------------------');
console.log('// Stats of Merge Sort //');
console.log('Array Length --> ', arr.length);
console.log('Time taken:', timeTakenInSeconds, 'seconds');

File: PYTHON/Algorithms/Sort/5. MergeSort.py
from ListGeneration import listGenerator
import time

arr50K = listGenerator(50000)

def merge(arr1, arr2):
    result = []
    i = 0
    j = 0
    while i < len(arr1) and j < len(arr2):
        if arr1[i] < arr2[j]:
            result.append(arr1[i])
            i += 1
        else:
            result.append(arr2[j])
            j += 1

    result.extend(arr1[i:])
    result.extend(arr2[j:])

    return result

def merge_sort(arr):
    if len(arr) < 2:
        return arr

    middle_index = len(arr) // 2
    left = merge_sort(arr[:middle_index])
    right = merge_sort(arr[middle_index:])

    return merge(left, right)

# Record the start time
start_time = time.time()

sorted_arr = merge_sort(arr50K)

# Record the end time
end_time = time.time()

# It will take at least 0.1 seconds to sort the arr50K

# if you want to visualize starting array just change arr to arr50K
print(sorted_arr)
print('----------------------------------------------------------------')
print('// Stats of Merge Sort //')
print('Length --> ', len(sorted_arr))
print("Time taken:", end_time - start_time, "seconds")